### PR TITLE
DROOLS-4433 : Add the ability to move constraints within patterns in the same way we can move patterns within rules to Guided Rules and Guided Rule Templates

### PR DESCRIPTION
**Pull request by Rikkola**
https://issues.jboss.org/browse/DROOLS-4433

https://github.com/kiegroup/drools/pull/2502
https://github.com/kiegroup/drools-wb/pull/1215

### CODE REVIEW

**Comment by Rikkola** [comment]
@manstis @jomarko If you two could please review.

**Review by jomarko — COMMENTED** [review]
I have no comments to code and manual check is pending.
However my attemtp to increase the coverage failed. @Rikkola may you please have a look if we have real issue or just my tests are wrong. Thank you.

https://github.com/Rikkola/drools/pull/4

**Comment by jomarko** [comment]
Manual review finished. Feedback provided in the second part of the ensemble.

File: drools-workbench-models/drools-workbench-models-datamodel-api/src/test/java/org/drools/workbench/models/datamodel/rule/util/FieldConstraintArrayUtilTest.java [hunk 1]
```diff
@@ -106,4 +106,53 @@ public void shouldNotMoveLowerThanUsedVariables() {
...
+        final SingleFieldConstraint top = new SingleFieldConstraint();
+        top.setConstraintValueType(BaseSingleFieldConstraint.TYPE_VARIABLE);
+        top.setValue("b");
+        fieldConstraints[0] = top;
+        final SingleFieldConstraint middle = new SingleFieldConstraint();
+        middle.setConstraintValueType(BaseSingleFieldConstraint.TYPE_VARIABLE);
+        middle.setValue("a");
+        fieldConstraints[1] = middle;
+        final SingleFieldConstraint bottom = new SingleFieldConstraint();
+        bottom.setFieldBinding("a");
```
Comment (Rikkola) [inline drools-workbench-models/drools-workbench-models-datamodel-api/src/test/java/org/drools/workbench/models/datamodel/rule/util/FieldConstraintArrayUtilTest.java:124]: Declaration should never be below the `SingleFieldConstraint `where it is used, but here our perfect GRE made a mistake, but it is possible to fix this with the move.

File: drools-workbench-models/drools-workbench-models-datamodel-api/src/test/java/org/drools/workbench/models/datamodel/rule/util/FieldConstraintArrayUtilTest.java [hunk 2]
```diff
@@ -106,4 +106,53 @@ public void shouldNotMoveLowerThanUsedVariables() {
...
+        bottom.setConstraintValueType(BaseSingleFieldConstraint.TYPE_VARIABLE);
+        bottom.setValue("b");
+        fieldConstraints[2] = bottom;
+
+        FieldConstraintArrayUtil.moveDown(0, fieldConstraints);
+        FieldConstraintArrayUtil.moveDown(1, fieldConstraints);
+
+        assertEquals(top, fieldConstraints[0]);
+        assertEquals(bottom, fieldConstraints[1]);
+        assertEquals(middle, fieldConstraints[2]);
```
Comment (Rikkola) [inline drools-workbench-models/drools-workbench-models-datamodel-api/src/test/java/org/drools/workbench/models/datamodel/rule/util/FieldConstraintArrayUtilTest.java:156]: Here `middle `can move freely, but `bottom `can not go over `top `and `top `can not go under `bottom`.

**Review by jomarko — APPROVED** [review]
Approving, please just merge my last coverage extension:
https://github.com/Rikkola/drools/pull/5

**Comment by Rikkola** [comment]
"org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M1:enforce (ban-duplicated-classes) on project kie-server-integ-tests-common: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed."

jenkins please compile downstream build

**Review by manstis — APPROVED** [review]

**Comment by Rikkola** [comment]
Pretty sure the build has been stuck for 3 days :)
jenkins please compile downstream build

**Pull request merged**